plotly/tools.py

```diff
@@ -2871,3 +2934,188 @@ def get_candle_decrease(self):
+        
+        dd_traces, xvals, yvals,
+        ordered_labels, leaves = self.get_dendrogram_traces(X, colorscale)
+        
+        self.labels = ordered_labels
+        self.leaves = leaves
+        yvals_flat = yvals.flatten()
+        xvals_flat = xvals.flatten()
+
+        self.zero_vals = []
+
+        for i in range(len(yvals_flat)):
+            if yvals_flat[i] == 0.0 and xvals_flat[i] not in self.zero_vals:
+                self.zero_vals.append(xvals_flat[i])
+
+        self.zero_vals.sort()
+                
+        self.layout = self.set_figure_layout(width, height)
+        self.data = Data(dd_traces)
+
+    def get_color_dict(self, colorscale):
+        """ 
+            Returns colorscale used for dendrogram tree clusters
+            :param (list) colorscale: colors to use for the plot,
+            in rgb format
+        """
+       
+        # These are the color codes returned for dendrograms
+        # We're replacing them with nicer colors
+        d = {'r': 'red',
+             'g': 'green',
+             'b': 'blue',
+             'c': 'cyan',
+             'm': 'magenta',
+             'y': 'yellow',
+             'k': 'black',
+             'w': 'white'}
+        default_colors = OrderedDict(sorted(d.items(), key=lambda t: t[0]))
+
+        if colorscale is None:
+            colorscale = [
+                "rgb(0,116,217)",   # blue
+                "rgb(35,205,205)",  # cyan
+                "rgb(61,153,112)",  # green
+                "rgb(40,35,35)",    # black
+                "rgb(133,20,75)",   # magenta
+                "rgb(255,65,54)",   # red
+                "rgb(255,255,255)",   # white
+                "rgb(255,220,0)"]   # yellow
+
+        for i in range(len(default_colors.keys())):
+            k = default_colors.keys()[i]
+            if i < len(colorscale):
+                default_colors[k] = colorscale[i]
+                
+        return default_colors
+   
+    def set_axis_layout(self, axis_key):
+        """ 
+            Sets and returns default axis object for dendrogram figure
+            :param (str) axis_key: "xaxis", "xaxis1", "yaxis", yaxis1", etc.
+        """
+
+        axis_defaults = {
+                'type': 'linear',
+                'ticks': 'outside',
+                'mirror': 'allticks',
+                'rangemode': 'tozero',
+                'showticklabels': True,                 
+                'zeroline': False,
+                'showgrid': False,              
+                'showline': True,
+            }
+        
+        if len(self.labels) != 0:
+            axis_key_labels = self.xaxis
+            if self.orientation in ['left', 'right']:
+                axis_key_labels = self.yaxis
+            if axis_key_labels not in self.layout:
+                self.layout[axis_key_labels] = {}
+            self.layout[axis_key_labels]['tickvals'] = [ea*self.sign[axis_key]
+                                                        for ea in self.zero_vals]
```

Comment:
:cow2: I don't understand what the `ea` variable means here? Just a little tough to read.